packages/angular_devkit/build_webpack/src/utils/webpack-file-system-host-adapter.ts

```diff
@@ -0,0 +1,154 @@
+import { map, mergeMap, switchMap } from 'rxjs/operators';
+
+
+export class WebpackFileSystemHostAdapter implements InputFileSystem {
+  protected _syncHost: virtualFs.SyncDelegateHost<Stats> | null = null;
+
+  constructor(protected _host: virtualFs.Host<Stats>) {}
+
+  private _doHostCall<T>(o: Observable<T>, callback: Callback<T>) {
+    let value: T | null = null;
+    let error = false;
+
+    try {
+      o.subscribe({
+        error(err) {
+          error = true;
+          callback(err);
+        },
+        next(v) {
+          value = v;
+        },
+        complete() {
+          if (value) {
+            callback(null, value);
+          } else {
+            callback(new Error('Unknown error happened.'));
+          }
+        },
+      });
+    } catch (err) {
+      // In some occasions, the error handler above will be called, then an exception will be
+      // thrown (by design in observable constructors in RxJS 5). Don't call the callback
+      // twice.
+      if (!error) {
+        callback(err);
+      }
+    }
+  }
+
+  stat(path: string, callback: Callback<virtualFs.Stats>): void {
+    const p = normalize('/' + path);
+    const result = this._host.stat(p);
+
+    if (result === null) {
+      const o = this._host.exists(p).pipe(
+        switchMap(exists => {
+          if (!exists) {
+            throw new FileDoesNotExistException(p);
+          }
+
+          return this._host.isDirectory(p).pipe(
+            mergeMap(isDirectory => {
+              return (isDirectory ? of(0) : this._host.read(p).pipe(
+                map(content => content.byteLength),
+              )).pipe(
+                map(size => [isDirectory, size]),
+              );
+            }),
+          );
+        }),
+        map(([isDirectory, size]) => {
+          return {
+            isFile() { return !isDirectory; },
+            isDirectory() { return isDirectory; },
+            size,
+            atime: new Date(),
+            mtime: new Date(),
+            ctime: new Date(),
+            birthtime: new Date(),
+          };
+        }),
+      );
+
+      this._doHostCall(o, callback);
+    } else {
+      this._doHostCall(result, callback);
+    }
+  }
+
+  readdir(path: string, callback: Callback<string[]>): void {
+    return this._doHostCall(this._host.list(normalize('/' + path)), callback);
+  }
+
+  readFile(path: string, callback: Callback<Buffer>): void {
+    const o = this._host.read(normalize('/' + path)).pipe(
+      map(content => new Buffer(content)),
+    );
+
+    return this._doHostCall(o, callback);
+  }
+
+  readJson(path: string, callback: Callback<JsonObject>): void {
+    const o = this._host.read(normalize('/' + path)).pipe(
+      map(content => JSON.parse(virtualFs.fileBufferToString(content))),
+    );
+
+    return this._doHostCall(o, callback);
+  }
+
+  readlink(path: string, callback: Callback<string>): void {
+    const err: NodeJS.ErrnoException = new Error('Not a symlink.');
```

Comment:
We should support symlinks in the future.

Reply:
I disagree, at least as far as the host is concerned, symlinks were intentionally left out of the design.

We "support" them by having the Node*Host supports proper stats and you check them, but if you don't know which host comes in then you cannot know (and shouldn't IMO).

Reply:
Ok, I don't have any usecase that needs them right now either.